### PR TITLE
[Test Framework] Add transaction execution support

### PR DESCRIPTION
**Pull request by SupunS**
Work towards #331

## Description

Add support for running transactions against the blockchain instance.

Also allows creating accounts, which would need to sign and authorize transactions. 

______

- [ ] Targeted PR against `master` branch
- [x] Linked to Github issue with discussion and accepted design OR link to spec that describes this work
- [x] Code follows the [standards mentioned here](https://github.com/onflow/cadence/blob/master/CONTRIBUTING.md#styleguides)
- [x] Updated relevant documentation 
- [x] Re-reviewed `Files changed` in the Github PR explorer
- [x] Added appropriate labels

### CODE REVIEW

File: runtime/stdlib/contracts/test.cdc [hunk 1]
```diff
@@ -77,16 +103,46 @@ pub contract Test {
...
+    // Account represents a info about the account created on the blockchain.
+    //
+    pub struct Account {
+        pub let address:   Address
+        pub let publicKey: PublicKey
+
+        init(_ address: Address, _ publicKey: PublicKey) {
+            self.address = address
+            self.publicKey = publicKey
+        }
```
Comment (turbolent) [inline runtime/stdlib/contracts/test.cdc:115]:
```suggestion
        pub let address: Address
        pub let publicKey: PublicKey

        init(address: Address, publicKey: PublicKey) {
            self.address = address
            self.publicKey = publicKey
        }
```

File: runtime/stdlib/contracts/test.cdc [hunk 2]
```diff
@@ -77,16 +103,46 @@ pub contract Test {
...
+        pub let authorizers: [Address]
+        pub let signers:     [Account]
+        pub let args:        [AnyStruct]
+
+        init(_ code: String, _ authorizers: [Address], _ signers: [Account], _ args: [AnyStruct]) {
+            self.code = code
+            self.authorizers = authorizers
+            self.signers = signers
+            self.args = args
+        }
```
Comment (turbolent) [inline runtime/stdlib/contracts/test.cdc:131]:
```suggestion
        pub let code: String
        pub let authorizers: [Address]
        pub let signers: [Account]
        pub let arguments: [AnyStruct]

        init(code: String, authorizers: [Address], signers: [Account], arguments: [AnyStruct]) {
            self.code = code
            self.authorizers = authorizers
            self.signers = signers
            self.arguments = arguments
        }
```

File: runtime/stdlib/contracts/test.cdc [hunk 3]
```diff
@@ -67,8 +91,10 @@ pub contract Test {
...
     }
 
+    // The result of a script execution.
+    //
     pub struct ScriptResult {
-        pub let status: ResultStatus
+        pub let status:      ResultStatus
         pub let returnValue: AnyStruct?
 
         init(_ status: ResultStatus, _ returnValue: AnyStruct?) {
```
Comment (turbolent) [inline runtime/stdlib/contracts/test.cdc:100]:
```suggestion
        init(status: ResultStatus, returnValue: AnyStruct?) {
```

File: runtime/stdlib/contracts/test.cdc [hunk 4]
```diff
@@ -67,8 +91,10 @@ pub contract Test {
         }
     }
 
+    // The result of a script execution.
+    //
     pub struct ScriptResult {
-        pub let status: ResultStatus
+        pub let status:      ResultStatus
```
Comment (turbolent) [inline runtime/stdlib/contracts/test.cdc:97]:
```suggestion
        pub let status: ResultStatus
```

File: runtime/stdlib/contracts/test.cdc [hunk 5]
```diff
@@ -77,16 +103,46 @@ pub contract Test {
...
+            self.args = args
+        }
+    }
+
     // BlockchainBackend is the interface to be implemented by the backend providers.
     //
     pub struct interface BlockchainBackend {
 
-        // fun addTransaction(_ transaction: Transaction)
+        pub fun executeScript(_ script: String, _ args: [AnyStruct]): ScriptResult
```
Comment (turbolent) [inline runtime/stdlib/contracts/test.cdc:138]:
```suggestion
        pub fun executeScript(_ script: String, _ arguments: [AnyStruct]): ScriptResult
```

File: runtime/interpreter/test-framework.go [hunk 6]
```diff
@@ -22,22 +22,51 @@
...
-	RunScript(code string, args []Value) ScriptResult
+	RunScript(code string, args []Value) *ScriptResult
+
+	CreateAccount() (*Account, error)
+
+	AddTransaction(
+		code string,
+		authorizers []common.Address,
+		signers []*Account,
+		args []Value,
```
Comment (turbolent) [inline runtime/interpreter/test-framework.go:43]: (sorry for being nitpicky about style)

```suggestion
	RunScript(code string, arguments []Value) *ScriptResult

	CreateAccount() (*Account, error)

	AddTransaction(
		code string,
		authorizers []common.Address,
		signers []*Account,
		arguments []Value,
```

File: runtime/stdlib/contracts/test.cdc [hunk 7]
```diff
@@ -12,53 +12,77 @@ pub contract Test {
             self.backend = backend
         }
 
+        // Executes a script and returns the script return value and the status.
+        // `returnValue` field of the result will be `nil` if the script failed.
+        //
         pub fun executeScript(_ script: String, _ args: [AnyStruct]): ScriptResult {
             return self.backend.executeScript(script, args)
```
Comment (turbolent) [inline runtime/stdlib/contracts/test.cdc:19]:
```suggestion
        pub fun executeScript(_ script: String, _ arguments: [AnyStruct]): ScriptResult {
            return self.backend.executeScript(script, arguments)
```

File: runtime/stdlib/test.go [hunk 8]
```diff
@@ -453,6 +497,455 @@ func createScriptResult(inter *interpreter.Interpreter, returnValue interpreter.
...
+	}
+
+	hashAlgoRawValue, ok := rawValue.(interpreter.UInt8Value)
+	if !ok {
+		panic(errors.NewUnreachableError())
+	}
+	return sema.HashAlgorithm(hashAlgoRawValue)
+}
+
+func publicKeyFromValue(
```
Comment (turbolent) [inline runtime/stdlib/test.go:766]: Looks like a duplicate of `runtime.NewPublicKeyFromValue`. Maybe move to e.g. `stdlib` (?). I think in one of my optimization/refactor PRs I also moved it

File: runtime/stdlib/test.go [hunk 9]
```diff
@@ -453,6 +497,455 @@ func createScriptResult(inter *interpreter.Interpreter, returnValue interpreter.
...
+			PublicKey: publicKey,
+		})
+
+		return true
+	})
+
+	return accounts
+}
+
+func hashAlgoFromValue(inter *interpreter.Interpreter, hashAlgoField interpreter.Value) sema.HashAlgorithm {
```
Comment (turbolent) [inline runtime/stdlib/test.go:748]: Looks like a duplicate of `runtime.NewHashAlgorithmFromValue`. Maybe move to e.g. `stdlib` and have `runtime` use it instead.

File: test-framework/emulator_backend.go [hunk 10]
```diff
@@ -39,20 +42,33 @@ var _ interpreter.TestFramework = &EmulatorBackend{}
 //
 type EmulatorBackend struct {
 	blockchain *emulator.Blockchain
+
+	// blockOffset is the offset for the sequence number of the next transaction.
+	// This is equal to the number of transactions in the current block.
+	// Must be rest once the block is committed.
```
Comment (turbolent) [inline test-framework/emulator_backend.go:48]:
```suggestion
	// Must be reset once the block is committed.
```

File: test-framework/test_framework_test.go [hunk 11]
```diff
@@ -387,3 +387,474 @@ func TestUsingEnv(t *testing.T) {
...
+
+                let tx = Test.Transaction(
+                    "transaction { execute{ assert(false) } }",
+                    [account.address],
+                    [account],
+                    [],
+                )
+
+                blockchain.addTransaction(tx)
+            }
```
Comment (turbolent) [inline test-framework/test_framework_test.go:429]: 😍

**Review by turbolent — COMMENTED** [review]
Great work!

File: runtime/stdlib/test.go (hunk 9 above)
Comment (SupunS) [reply runtime/stdlib/test.go]: This is actually no longer needed. This was a left-over I missed out to clean up.

**Review by turbolent — APPROVED** [review]
Nice work!

**Pull request merged**